### PR TITLE
[QE] Continual tests for KafkaSource for upgrade testing

### LINKED ISSUE #67: Upgrade tests for KafkaSource
We need to implement upgrade tests for KafkaSource and KafkaBinding so that in the future we can change the storage version to v1beta1 from v1alpha1.

Best idea is to keep eventing at latest release and only changing the eventing-kafka version. We can in the future run the tests again with eventing head too.

Part of umbrella issue, #65

### PR DESCRIPTION
**Pull request by cardil**
Fixes #67 

## Proposed Changes

- Adds an implementation of `KafkaSource` continual testing by sending events to Kafka topic directly

## Additional notes

This PR requires merge of the following PRs:

 - knative/eventing#4815
 - knative/eventing#5321

### CODE REVIEW

**Comment by cardil** [comment]
This PR requires merge of the following PRs:

- knative/eventing#4815
- TBD

File: go.mod
Comment (cardil) [inline go.mod:44]: /hold

This PR uses integration branch of https://github.com/cardil/eventing/tree/integration/eventing-sandbox%2Feventing-kafka%2Fpr-563 which merges knative/eventing#4815 and knative/eventing#5321.

Do not merge this PR, until those PRs are merged to eventing and this code is restored.

File: test/upgrade/continual.go
Comment (mgencur) [inline test/upgrade/continual.go:26]: Nit: `event` -> `events`

File: test/upgrade/continual/kafkasender.go
Comment (mgencur) [inline test/upgrade/continual/kafkasender.go:38]: Nit: The words `given` is probably redundant here.

File: test/upgrade/continual/options.go
Comment (mgencur) [inline test/upgrade/continual/options.go:23]: Maybe these should be `const` ? They're not being re-set anywhere so it looks like they're consts.

File: test/upgrade/continual/source.go
Comment (mgencur) [inline test/upgrade/continual/source.go:93]: It seems like it would have been better to fill in the defaults when creating the kafka cluster via `fillInDefaults` and in this place only return `c.Name`` instead of returning some default value which might not be correct. Just shifting the logic a little bit.

File: test/upgrade/continual/source.go
Comment (mgencur) [inline test/upgrade/continual/source.go:101]: Same comment as above.

**Review by mgencur — COMMENTED** [review]
Hey @cardil , I think it looks great. A few comments below but they're mostly nits.

File: go.mod
Comment (cardil) [reply go.mod]: knative/eventing#4815 and knative/eventing#5321 are merged and integration branch is removed.

/unhold

File: test/upgrade/continual/source.go
Comment (cardil) [reply test/upgrade/continual/source.go:93]: Hmm, maybe. I'll take a look...

Comment (cardil) [reply test/upgrade/continual/source.go]: I've made a refactor on how to pass those variables and set default values. Now it's also possible to influence retry and replication options.

Changes are in this commit: https://github.com/knative-sandbox/eventing-kafka/pull/563/commits/5fa1ec780d22cc4b4416db84727b87250cfce709

**Comment by cardil** [comment]
This lint error: https://github.com/knative-sandbox/eventing-kafka/pull/563/checks?check_run_id=2658845322 is due to https://github.com/knative-sandbox/eventing-kafka/pull/648

File: test/upgrade/suite.go
Comment (mgencur) [inline test/upgrade/suite.go:61]: Hopefully, we won't get more than two slices :)
I'm curious if this could be even more general. I searched this code via Google but it could be something like it (just changing types...):
```
func mergeSlices(args...[]int) []int{
	mergedSlice := make([]int, 0)
	for _, oneSlice := range  args {
		mergedSlice = append(mergedSlice, oneSlice...)
	}
	return mergedSlice
}
```

**Review by mgencur — COMMENTED** [review]
One minor suggestion about joining slices. Otherwise looks good.

File: test/upgrade/continual/source.go
Comment (mgencur) [reply test/upgrade/continual/source.go]: Nice!

File: test/upgrade/suite.go
Comment (cardil) [reply test/upgrade/suite.go:61]: Downside of this implementation is allocation of new memory segment for each loop as capacity for slide is calculated automatically as `len(x)*2` - that's true for small slices.

Maybe this impl. would be better:

```golang
func merge(slices... []pkgupgrade.BackgroundOperation) []pkgupgrade.BackgroundOperation {
	l := 0
	for _, slice := range slices {
		l += len(slice)
	}
	result := make([]pkgupgrade.BackgroundOperation, 0, l)
	for _, slice := range slices {
		result = append(result, slice...)
	}
	return result
}
```

Comment (mgencur) [reply test/upgrade/suite.go:61]: Yeah. That looks good.

File: test/upgrade/continual/channel-config.toml
Comment (matzew) [inline test/upgrade/continual/channel-config.toml:4]: why is the last char here _red_ ?

File: test/upgrade/continual/source-config.toml
Comment (matzew) [inline test/upgrade/continual/source-config.toml:4]: is this formatted correctly ?

File: test/upgrade/continual/channel-config.toml
Comment (cardil) [reply test/upgrade/continual/channel-config.toml:4]: This is because this is template of a toml, but file is names `*.toml`. Github thinks this is regular TOML, and then it's invalid.

The `interval = {{ .Config.Interval.Nanoseconds }}` is replaced and actual TOML file is something like: `interval = 10000000`, and that's valid TOML.

File: test/upgrade/continual/channel.go
Comment (cardil) [reply test/upgrade/continual/channel.go:46]: Actually, I had this as an imported constant before. At some point you've replaced that with string. See: https://github.com/knative-sandbox/eventing-kafka/blame/main/test/upgrade/test_main.go#L33

Comment (cardil) [reply test/upgrade/continual/channel.go:46]: That was in https://github.com/knative-sandbox/eventing-kafka/pull/513

File: test/upgrade/continual/source-config.toml
Comment (cardil) [reply test/upgrade/continual/source-config.toml:4]: Yes. It's advised to indent the part if you are setting some key as object. Here, I'm setting address with `{bootstrapServers: '', topicName: ''}`.

See example: https://github.com/toml-lang/toml#example

File: test/upgrade/continual/channel.go
Comment (cardil) [reply test/upgrade/continual/channel.go:46]: I even asked about that: https://github.com/knative-sandbox/eventing-kafka/pull/513/files#r608632600 :rofl:

Comment (matzew) [reply test/upgrade/continual/channel.go:46]: OK!  Fair point, I will send a PR 

thanks :-)

File: test/upgrade/continual/source-config.toml
Comment (matzew) [reply test/upgrade/continual/source-config.toml:4]: weird :-)

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
